### DISCUSSION TITLE
[help] Debugging long-running functions applied to many groups

### DISCUSSION THREAD

**Discussion by eliocamp** [ropensci/targets#1224] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

One of my targets is a preprocessing step that takes a long time and results in a large quantity of files an branches over many groups. Something like this:

```r
     , tar_target(preprocessed_files, preprocess(files),
                  pattern = map(files), 
                  format = "file_fast"
                  )
```

When some branches error out and I have to debug, I often encounter subtle edge cases that only appear in some groups and I need to modify `preprocess`'s code to accommodate them. Unfortunately, this changes the signature of the function, invalidates all my previous branches and all of them need to be run again. 

This is reasonable because even though I know (or suspect) that the changes only affect some groups, target doesn't know that and I might be mistaken, so this is the correct behaviour for reproducibility. 

So my question is, are there any techniques to make this easier and faster? 

I was thinking that one approach could be to have a first target that takes the group and then returns a list of functions to apply to the group --maybe add that list as an attribute of the group. Then the preprocess step would be a function that just applies those functions one after the other. That way, only groups with changed attributes or whose attributes includes functions that changed would be re-run. Does that make any sense?

**Comment by wlandau** [comment]
I would have a look at the ‘debug’ option in tar_option_set(), which can accept the names of individual (errored) branches: https://books.ropensci.org/targets/debugging.html#pause-the-pipeline-with-the-debug-option. The whole chapter is worth a read for your case.
